CONTRIBUTING.md

```diff
@@ -0,0 +1,233 @@
+  //  2. Ensures that folders that have been deleted by git 
+  //     (but are still in the projection) are found before their
+  //     parent folder is re-expanded (only applies on platforms 
+  //     where EnumerationExpandsDirectories is true)
+  foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in   placeholderFoldersListCopy.OrderByDescending(x => x.Path))
+  ```
+
+  Obvious (bad) comment:
+
+  ```
+  // Check if enumeration expands directories on the current platform
+  if (GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories)
+  ```
+
+- *Add new interfaces when it makes sense for the product, not simply for unit testing*
+
+  When a class needs to be mocked (or have a subset of its behavior mocked), prefer using virtual methods instead of adding a new interface.  VFS4G uses interfaces when multiple implementations of the interface exist in the product code.
+
+- *Check for `null` using the equality (`==`) and inequality (`!=`) operators rather than `is`*
+
+  A corollary to this guideline is that equality/inequality operators that break `null` checks should not be added (see [this post](https://stackoverflow.com/questions/40676426/what-is-the-difference-between-x-is-null-and-x-null) for an example).
+
+- *Use `nameof(...)` rather than hardcoded strings*
+
+  Using `nameof` ensures that when methods/variables are renamed the logging of those method/variable names will also be updated.  However, hard coded strings are still appropriate when they are used for generating reports and changing the strings would break the reports.
+
+### C/C++
+
+- *Do not use C-style casts.  Use C++-style casts.*
+
+  C++ style casts (e.g. `static_cast<T>`) more clearly express the intent of the programmer, allow for better validation by the compiler, and are easier to search for in the codebase.
+
+- *Declare static functions at the top of source files*
+  
+  This ensures that the functions can be called from anywhere inside the file.
+
+- *Do not use namespace `using` statements in header files*
+
+  `using` statements inside header files are picked up by all source files that include the headers and can cause unexpected errors if there are name collisions.
+
+- *Prefer `using` to full namespaces in source files*
+
+  Example:
+  ```
+  // Inside MyFavSourceFile.cpp
+
+  using std::string;
+
+  // Do not use `std::` namespace with `string`
+  static string s_myString;
+  ```
+
+- *Use a meaningful prefix for "public" free functions, and use the same prefix for all functions in a given header file*
+
+   Example:
+   ```
+   // Functions declared in VirtualizationRoots.h have "VirtualizationRoot_" prefix
+   bool VirtualizationRoot_IsOnline(VirtualizationRootHandle rootHandle);
+
+   // Static helper function in VirtualizationRoots.cpp has no prefix
+   static VirtualizationRootHandle FindOrDetectRootAtVnode(vnode_t vnode, const FsidInode& vnodeFsidInode);
+   ```
+
```

Comment:
We had a slightly longer, more comprehensive list for C++ style guidelines on a wiki somewhere prior to the move to GitHub. I won't claim to know where that ended up…

Reply:
@pmj thanks for the tip, I'll see if I can dig those up.

Reply:
@pmj I believe I've found the style guide you mentioned in a OneNote.  

I've filed #894 to move those guidelines over here and will do that in a follow-up PR.